python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+            sleep(0.1)
+
+        ### Runner Pool
+        # Wait for the runner pool to empty before waiting on the status pool
+        # The runner pool and status pool can exhange jobs between each other but it is always the status pool
+        # that will have items in the queue last (that last job with a finished status to be printed)
+
+        # Close the runner_pool so jobs can no longer be scheduled for launching
+        self.runner_pool.close()
+
+        # Wait for the runner_pool to empty
+        self.runner_pool.join()
+
+        # Close the status_pool so jobs can no longer be scheduled for status printing
+        self.status_pool.close()
+
+        # Wait for the status_pool to empty
+        self.status_pool.join()
+
+    # Method to do things needed to actively running tests.
+    # NOTE: We are in a thread locked state, so be quick
+    def handleActiveTests(self):
+        # loop through all the active tests and do something
+        for tester_name in self.active_jobs:
+            tester = self.scheduled_name_to_object[tester_name]
+
+            # Handle long running tests
+            self.doLongRunningTests(tester)
+
+            # Handle timeouts
+            self.doTimeoutTests(tester)
+
+    def doLongRunningTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            min_time = self.default_reported_time
+            if tester.specs.isValid('min_reported_time'):
+                min_time = float(tester.getMinReportTime())
+
+            if clock() - tester.getStartTime() > min_time:
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('RUNNING...', tester.bucket_pending)
+                self.harness.handleTestStatus(tester)
+
+    def doTimeoutTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            if clock() - tester.getStartTime() > float(tester.getMaxTime()):
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('TIMEOUT', tester.bucket_fail)
+                self.harness.handleTestStatus(tester)
+
+    # TODO: Possibly look into semaphore functionality for slot allocation
+    def checkAvailableSlots(self, tester):
+        # See if we can fit this job into our busy schedule
+        if self.workers_in_use + tester.getProcs(self.options) <= self.job_slots:
+            return True
+
+        # Check for insufficient slots -soft limit
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and self.soft_limit:
+            tester.specs.addParam('caveats', ['OVERSIZED'], "")
+            return True
+
+        # Check for insufficient slots -hard limit (skip this job)
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and not self.soft_limit:
+            tester.setStatus('insufficient slots', tester.bucket_skip)
```

Comment:
Both of these checks can happen before you get to the threaded region. Then you can pull the dynamic check out and just put the logic directly in where you are using it right under the thread lock.